### PR TITLE
Use KDE for beam screen images

### PR DESCRIPTION
**Pull request by cr-xu**
## Description

Add an option for `Screen` element. Now it can use KDE to get the beam image which allows differentiation.
Behavior can be switched using `method="histogram"` and `method="kde"`.
It contains a new feature `kde_bandwidth` for smoothing the beam image.

The returned image has the convention [...,Y, X] where y-axis starts from the top of the image, e.g. `plt.imshow()` will show the image upright

```python

import cheetah
import torch
import matplotlib.pyplot as plt

incoming = cheetah.ParticleBeam.from_parameters(num_particles = 10000, mu_x = torch.tensor([-0.5e-3]), mu_y=torch.tensor([1.0e-3]), sigma_x = torch.tensor([1e-4]), sigma_y = torch.tensor([3e-4]))

screen = cheetah.Screen(resolution=(1200, 800), pixel_size=(5e-6, 5e-6), binning=torch.tensor(2), is_active=True)

screen.track(incoming)

plt.imshow(screen.reading[0], extent=screen.extent, cmap='gray')
plt.xlabel('x [m]')
plt.ylabel('y [m]')
```

image

## Motivation and Context

- [x] I have raised an issue to propose this change (required for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (_required for a bug fix or a new feature_).
- [ ] I have updated the documentation accordingly.
  
  
- [x] I have reformatted the code and checked that formatting passes (**required**).
- [x] I have have fixed all issues found by `flake8` (**required**).
- [x] I have ensured that all `pytest` tests pass (**required**).
- [ ] I have run `pytest` on a machine with a CUDA GPU and made sure all tests pass (**required**).
- [x] I have checked that the documentation builds (**required**).

Note: We are using a maximum length of 88 characters per line

### CODE REVIEW

**Comment by cr-xu** [comment]
Hmm... so KDE clearly makes the screen reading slower, as expected :)

@jank324 Shall we 
- Change the test file to make the speed limit more generous?
- Still provide a basic histogramdd() reading for normal tracking, when differentiation w.r.t. images are not needed? This can be switched e.g. with `use_kde=False` by default?

**Comment by jank324** [comment]
I think we should provide both. Im just wondering whether it should be `use_kde=True` or `method=`kde`.

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,7 @@
 - Add Python 3.12 support (see #161) (@jank324)
 - Implement space charge using Green's function in a `SpaceChargeKick` element (see #142) (@greglenerd, @RemiLehe, @ax3l, @cr-xu, @jank324)
 - `Segment`s can now be imported from Bmad to devices other than `torch.device("cpu")` and dtypes other than `torch.float32` (see #196, #206) (@jank324)
+- `Screen` will now use KDE for differentiable images. (see #200) (@cr-xu, @roussel-ryan
```
Comment (jank324) [inline CHANGELOG.md:16]:
```suggestion
- `Screen` will now use KDE for differentiable images. (see #200) (@cr-xu, @roussel-ryan)
```

File: cheetah/accelerator/screen.py [hunk 2]
```diff
@@ -38,7 +43,9 @@ def __init__(
         pixel_size: Optional[Union[torch.Tensor, nn.Parameter]] = None,
         binning: Optional[Union[torch.Tensor, nn.Parameter]] = None,
         misalignment: Optional[Union[torch.Tensor, nn.Parameter]] = None,
+        kde_bandwith: Optional[Union[torch.Tensor, nn.Parameter]] = None,
         is_active: bool = False,
+        method: str = "histogram",
```
Comment (jank324) [inline cheetah/accelerator/screen.py:48]:
```suggestion
        method: Literal["histogram", "kde"] = "histogram",
```

File: cheetah/utils.py [hunk 3]
```diff
@@ -9,3 +18,170 @@ def __call__(self):
...
+
+    if not isinstance(bins, torch.Tensor):
+        raise TypeError(f"Input bins type is not a torch.Tensor. Got {type(bins)}")
+
+    if not isinstance(sigma, torch.Tensor):
+        raise TypeError(f"Input sigma type is not a torch.Tensor. Got {type(sigma)}")
+
+    if not bins.dim() == 1:
+        raise ValueError(
+            "Input bins must be a of the shape NUM_BINS" " Got {}".format(bins.shape)
```
Comment (jank324) [inline cheetah/utils.py:58]: We should use f-strings.

File: cheetah/utils.py [hunk 4]
```diff
@@ -9,3 +18,170 @@ def __call__(self):
...
+    Returns:
+        Computed histogram of shape :math:`(B, N_{bins}), N_{bins})`.
+
+    Examples:
+        >>> x1 = torch.rand(2, 32)
+        >>> x2 = torch.rand(2, 32)
+        >>> bins = torch.torch.linspace(0, 255, 128)
+        >>> hist = kde_histogram_2d(x1, x2, bins, bandwidth=torch.tensor(0.9))
+        >>> hist.shape
+        torch.Size([2, 128, 128])
```
Comment (jank324) [inline cheetah/utils.py:179]: I think this example doesn't actually match the modified version (?)

File: cheetah/utils.py [hunk 5]
```diff
@@ -9,3 +18,170 @@ def __call__(self):
...
+    return pdf
+
+
+def kde_histogram_2d(
+    x1: torch.Tensor,
+    x2: torch.Tensor,
+    bins1: torch.Tensor,
+    bins2: torch.Tensor,
+    bandwidth: torch.Tensor,
+    weights=None,
```
Comment (jank324) [inline cheetah/utils.py:155]: What type is `weights`?

**Comment by jank324** [comment]
@cr-xu could you have a look at the docstrings? Some of the returns just said what type the return is, but not what it is, and I wasn't sure.

**Comment by jank324** [comment]
Also, looking at this, would it make sense to split `utils.py` into
```
utils/__init__.py
utils/kde.py
utils/unique_name_generator.py
```
within this PR or do that later on?

**Comment by jank324** [comment]
And last one, @cr-xu was there a reason you chose this implementation over the one you had? Just wondering.

**Comment by cr-xu** [comment]
> And last one, @cr-xu was there a reason you chose this implementation over the one you had? Just wondering.

They are equivalent in math but I find this implementation just cleaner and more readable.

> @cr-xu could you have a look at the docstrings? Some of the returns just said what type the return is, but not what it is, and I wasn't sure.

I believe you fixed all the docstrings, right? Thanks!

**Comment by jank324** [comment]
> > And last one, @cr-xu was there a reason you chose this implementation over the one you had? Just wondering.
> 
> They are equivalent in math but I find this implementation just cleaner and more readable.
> 
> > @cr-xu could you have a look at the docstrings? Some of the returns just said what type the return is, but not what it is, and I wasn't sure.
> 
> I believe you fixed all the docstrings, right? Thanks!

I didn't fix every docstring ... just the ones I could easily figure out what the return is. So it would be good if you could look over if there are missing ones, and if what I put in makes sense.

File: CHANGELOG.md [hunk 6]
```diff
@@ -13,6 +13,7 @@
 - Add Python 3.12 support (see #161) (@jank324)
 - Implement space charge using Green's function in a `SpaceChargeKick` element (see #142) (@greglenerd, @RemiLehe, @ax3l, @cr-xu, @jank324)
 - `Segment`s can now be imported from Bmad to devices other than `torch.device("cpu")` and dtypes other than `torch.float32` (see #196, #206) (@jank324)
+- `Screen` will now use KDE for differentiable images. (see #200) (@cr-xu, @roussel-ryan)
```
Comment (jank324) [inline CHANGELOG.md:16]:
```suggestion
- `Screen` now offers the option to use KDE for differentiable images (see #200) (@cr-xu, @roussel-ryan)
```

**Comment by cr-xu** [comment]
> I didn't fix every docstring ... just the ones I could easily figure out what the return is. So it would be good if you could look over if there are missing ones, and if what I put in makes sense.

I looked over the code again and fixed a few things. Now the PR is good to go from my side. 
A potential TODO would be to add an example case in the documentation page, but that can be done in future.

File: tests/test_kde.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    )
+
+    hist_neglect_weights = kde_histogram_2d(
+        x_weighted[:, 0], x_weighted[:, 1], bins1, bins2, sigma
+    )
+    print(hist_unweighted[5])
+    print(hist_weighted[5])
+    print(hist_neglect_weights[5])
+    assert torch.allclose(hist_unweighted, hist_weighted)
+    assert not torch.allclose(hist_weighted, hist_neglect_weights)
```
Comment (jank324) [inline tests/test_kde.py:51]: @cr-xu could you add a description to these?

**Review by jank324 — APPROVED** [review]

**Comment by cr-xu** [comment]
Here's a quick study of the computation scaling for KDE and histogram methods:

The pixel counts are the total numbers in a screen (e.g. 1e6 corresponds to 1000*1000 screen).

image
